logictest: add a local-read-committed default config

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/100178
Release note: None

rafiss committed:
logictest: add a local-read-committed default config

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107990)
<!-- Reviewable:end -->

michae2 commented:
Thanks for creating this!

I guess basically every logictest does `CREATE TABLE`. 🤔

rafiss committed:
sql: fix autocommit_before_ddl to work for all multi-stmt txns

This fixes a bug where the setting would not work correctly if an
implicit transaction had multiple statements.

Release note: None

rafiss committed:
tree: mark DISCARD as a canModifySchema statement

DISCARD can result in temporary tables being dropped, so it should be
marked as a statement that can cause schema changes.

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/logictest/testdata/logic_test/cascade` line 3947 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NpzPOdzBB6v3SsZJe55:-NpzPOe-0ZaLyZxQMQrv:blw4wgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb10c4cfeecc1b0a3e3ec4ec6cafeb96d1ad71e6/pkg/sql/logictest/testdata/logic_test/cascade#L3947)):*
> ```
> statement ok
> BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;
> SET LOCAL autocommit_before_ddl = false;
> ```

to reviewers: the 3rd commit updates many logic test files in a mechanical way: in many situations where the test file had a multi-statement transaction, the transaction initialization was changed to explicitly use SERIALIZABLE and turn of the `autocommit_before_ddl` setting. this is because the schema changer has different behavior under multi-statement transactions, so rather than updating all the tests to account for the `autocommit_before_ddl` setting, it seemed easier to just force the test to use a multi-statement txn.

this makes the diff pretty large. let me know if you prefer that i break up this PR further in some way.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thank you for doing this! It is monumental!

I'm a little under halfway through. It's looking good.

So far, the only thing I'm finding difficult is the occasional `skipif config local-read-committed` on statements in the middle of explicit transactions that have presumably been affected by `autocommit_before_ddl`. The silent commit-on-DDL behavior is tricky, especially since there's not much in the test reminding the reader that the state of the txn might not match what they read.

What would you think of setting `autocommit_before_ddl` directly in tests, rather than for the config? You can say no, since that's probably much more painful... but it might make these tests easier to understand.

Or alternatively, what if we mostly used serializable transactions in these cases, and kept `autocommit_before_ddl` false, with only a few specific tests for `autocommit_before_ddl` rather than enabling it by default? That would reduce the need to `SET LOCAL autocommit_before_ddl = false;` everywhere and I think would be closer to what people expect.

> Reviewed 21 of 45 files at r3, 30 of 30 files at r4, 1 of 1 files at r5, 26 of 118 files at r6, 1 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-:-Nq05-4dCpKc2Zjf9e6k:b-q764db)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @rafiss)

___
*[`pkg/sql/conn_executor_exec.go` line 942 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq-kPdR6Nf0kNJ3xlB3:-Nq-kPdR6Nf0kNJ3xlB4:b-xxvilu) ([raw file](https://github.com/cockroachdb/cockroach/blob/865f0dbc58db2fd1814e46f1e49dbb1f69f436aa/pkg/sql/conn_executor_exec.go#L942)):*
> ```Go
> 	if tree.CanModifySchema(ast) &&
> 		((!ex.planner.EvalContext().TxnIsSingleStmt && ex.extraTxnState.firstStmtExecuted) ||
> 			!ex.implicitTxn()) &&
> ```

Should we also be checking `ex.extraTxnState.firstStmtExecuted` in the explicit transaction case?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor` line 814 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq-pkqG4g64QJVll8aw:-Nq-pkqG4g64QJVll8ax:bmigzss) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f36f1306d694ef02d80b609bf86e5aa485bfab3/pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor#L814)):*
> ```
> <unnamed portal 10>
> 
> skipif config local-read-committed
> ```

I think you mentioned this during the meeting, but why don't these cursors work, again? This first case seems to work when I try it by hand, at least.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 1566 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq00I_E1DY46lAVs9Od:-Nq00I_E1DY46lAVs9Oe:bkc8ypn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f36f1306d694ef02d80b609bf86e5aa485bfab3/pkg/sql/logictest/testdata/logic_test/alter_table#L1566)):*
> ```
> 
> # The unique constraint prevents new duplicate values.
> skipif config local-read-committed 110873
> ```

Thanks for linking to the issue!
___
*[`pkg/sql/logictest/testdata/logic_test/cascade` line 3947 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NpzPOdzBB6v3SsZJe55:-Nq-zuUl81yZC16NmBg9:b-w3fmtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb10c4cfeecc1b0a3e3ec4ec6cafeb96d1ad71e6/pkg/sql/logictest/testdata/logic_test/cascade#L3947)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

to reviewers: the 3rd commit updates many logic test files in a mechanical way: in many situations where the test file had a multi-statement transaction, the transaction initialization was changed to explicitly use SERIALIZABLE and turn of the `autocommit_before_ddl` setting. this is because the schema changer has different behavior under multi-statement transactions, so rather than updating all the tests to account for the `autocommit_before_ddl` setting, it seemed easier to just force the test to use a multi-statement txn.

this makes the diff pretty large. let me know if you prefer that i break up this PR further in some way.
</blockquote></details>

This is fine with me.
___
*[`pkg/sql/logictest/testdata/logic_test/check_constraints` line 152 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq00bH7CGbV64DGdwKv:-Nq00bH7CGbV64DGdwKw:burza23) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f36f1306d694ef02d80b609bf86e5aa485bfab3/pkg/sql/logictest/testdata/logic_test/check_constraints#L152)):*
> ```
> 
> statement ok
> UPDATE t4 SET a = 0 WHERE c = 3
> ```

Nice, thank you for trying it both ways.
___
*[`pkg/sql/logictest/testdata/logic_test/grant_in_txn` line 216 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq04Wnz6WFIC6XR0Tj3:-Nq04Wnz6WFIC6XR0Tj4:bwjai9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f36f1306d694ef02d80b609bf86e5aa485bfab3/pkg/sql/logictest/testdata/logic_test/grant_in_txn#L216)):*
> ```
> GRANT role_bar TO testuser;
> 
> skipif config local-read-committed
> ```

Why is this one needed?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
i think you're right. the silent autocommit behavior is tricky when it's not obvious that the setting was enabled.

> what if we mostly used serializable transactions in these cases, and kept `autocommit_before_ddl` false, with only a few specific tests for `autocommit_before_ddl` rather than enabling it by default? That would reduce the need to `SET LOCAL autocommit_before_ddl = false;` everywhere and I think would be closer to what people expect.

i will try this approach. it seems much easier to understand as a reader

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-:-Nq69D9xEldbBqbWXiJB:bhopw2y)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/conn_executor_exec.go` line 942 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq-kPdR6Nf0kNJ3xlB3:-Nq63XCy3nMazNHI7XVl:b-79hojr) ([raw file](https://github.com/cockroachdb/cockroach/blob/865f0dbc58db2fd1814e46f1e49dbb1f69f436aa/pkg/sql/conn_executor_exec.go#L942)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should we also be checking `ex.extraTxnState.firstStmtExecuted` in the explicit transaction case?
</blockquote></details>

that seems like a good change to make
___
*[`pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor` line 814 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq-pkqG4g64QJVll8aw:-Nq65fAh9jdsQCJE8J5N:bg4bjga) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f36f1306d694ef02d80b609bf86e5aa485bfab3/pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor#L814)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think you mentioned this during the meeting, but why don't these cursors work, again? This first case seems to work when I try it by hand, at least.
</blockquote></details>

i get this error:
```
plpgsql_cursor:815: SELECT name FROM pg_cursors;
        expected:
            <unnamed portal 10>
            foo

        but found (query options: "rowsort" -> ignore the following ordering of rows) :100: 

```

but i realize now this is just because the `CREATE OR REPLACE FUNCTION` made the transaction commit, and close the existing cursors. i can make this turn off the autocommit_before_ddl setting and use serializable, like in the other tests
___
*[`pkg/sql/logictest/testdata/logic_test/grant_in_txn` line 216 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq04Wnz6WFIC6XR0Tj3:-Nq64KZj-gZQKblQA8vv:b-74yoep) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f36f1306d694ef02d80b609bf86e5aa485bfab3/pkg/sql/logictest/testdata/logic_test/grant_in_txn#L216)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why is this one needed?
</blockquote></details>

this would fail with a "savepoint `before_invalid_grant` does not exist" error, since the transaction would have been auto-committed by this point.

aside: this is related to one of the reasons why i didn't think it was that beneficial to implement the autocommit_before_ddl setting as `COMMIT; DDL; BEGIN;` -- it feels potentially more confusing for the user to end up in a new explicit transaction, but none of the state from the earlier transaction state would carry over. it seemed easier to reason that the transaction was simply over due to the presence of DDL. (that's the mysql behavior too fwiw: https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html) 


<!-- Sent from Reviewable.io -->

rafiss commented:
This is ready for another look

michae2 reviewed (changes requested):
Thanks! That helped a lot.

> Reviewed 5 of 118 files at r6, 116 of 116 files at r8, 1 of 1 files at r9, 129 of 130 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-:-NqASag5EJkeBga1AsTQ:b-w6oyls)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @rafiss)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/schema_change_in_txn` line 26 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqARiWH43EWuW4xlRGa:-NqARiWH43EWuW4xlRGb:b-l3kfjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/ccl/logictestccl/testdata/logic_test/schema_change_in_txn#L26)):*
> ```
> 
> # Before the change which introduced this test, it would erroneously return 2.
> skipif config local-read-committed
> ```

Maybe this should be a serializable transaction instead?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/udf_plpgsql` line 2866 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqARzDM55S-iSM48BHY:-NqARzDM55S-iSM48BHZ:bzclyvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/ccl/logictestccl/testdata/logic_test/udf_plpgsql#L2866)):*
> ```
> $$ LANGUAGE PLpgSQL;
> 
> skipif config local-read-committed
> ```

Why is this one skipped?
___
*[`pkg/sql/logictest/testdata/logic_test/fk` line 4175 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqA8qGh2a_3aWOh5Mzc:-NqA8qGh2a_3aWOh5Mzd:b-mnjor9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/fk#L4175)):*
> ```
> 
> statement ok
> DELETE FROM parent80828 WHERE id = 1;
> ```

Does this table exist under RC?
___
*[`pkg/sql/logictest/testdata/logic_test/grant_in_txn` line 216 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-Nq04Wnz6WFIC6XR0Tj3:-NqAA-ob2dhZgpTGgJR_:bu40zb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f36f1306d694ef02d80b609bf86e5aa485bfab3/pkg/sql/logictest/testdata/logic_test/grant_in_txn#L216)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this would fail with a "savepoint `before_invalid_grant` does not exist" error, since the transaction would have been auto-committed by this point.

aside: this is related to one of the reasons why i didn't think it was that beneficial to implement the autocommit_before_ddl setting as `COMMIT; DDL; BEGIN;` -- it feels potentially more confusing for the user to end up in a new explicit transaction, but none of the state from the earlier transaction state would carry over. it seemed easier to reason that the transaction was simply over due to the presence of DDL. (that's the mysql behavior too fwiw: https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html)
</blockquote></details>

Ack. Thanks for explaining. It's kind of a "worse is better" argument, and I buy that.
___
*[`pkg/sql/logictest/testdata/logic_test/lock_timeout` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAFP590d2TncQl1Fyf:-NqAFP5A0VwKZ_zZaPI5:b8aa54v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/lock_timeout#L3)):*
> ```
> # LogicTest: !local-read-committed
> # The locks created in this test are intended to work with SERIALIZABLE
> # transactions.
> ```

What happens when this one is run under RC? Do lock timeouts not work with replicated locks?
___
*[`pkg/sql/logictest/testdata/logic_test/manual_retry` line 1 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAHC4F5JHP8ynlkvOv:-NqAHC4F5JHP8ynlkvOw:b-sbd3y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/manual_retry#L1)):*
> ```
> # LogicTest: !local-read-committed
> ```

Might be good to add a comment about automatically retrying under RC?
___
*[`pkg/sql/logictest/testdata/logic_test/read_committed` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAHwThEX2pViGI6Awu:-NqAHwThEX2pViGI6Awv:b8vtrhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/read_committed#L3)):*
> ```
> # LogicTest: !local-mixed-23.1 !local-read-committed
> # Don't run this test under local-read-committed, since that automatically
> # sets autocommit_before_ddl and default_transaction_isolation, which we
> ```

I guess the `autocommit_before_ddl` part is outdated now?
___
*[`pkg/sql/logictest/testdata/logic_test/schema_change_in_txn` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqALWR091PrcwZPwN8r:-NqALWR091PrcwZPwN8s:bb614rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn#L3)):*
> ```
> # LogicTest: !local-read-committed
> # This test is skipped under local-read-committed, since it uses the
> # autocommit_before_ddl setting, which makes it impossible to test schema
> ```

Outdated comment.
___
*[`pkg/sql/logictest/testdata/logic_test/show_commit_timestamp` line 331 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAPNe44qqd6ANQBiP4:-NqAPNe44qqd6ANQBiP5:bl6bffb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/show_commit_timestamp#L331)):*
> ```
> 
> # The timestamps are different under READ COMMITTED.
> skipif config local-read-committed
> ```

Hmm. Is `SHOW COMMIT TIMESTAMP` broken under RC? Oh, is it because the transaction isn't pushed? But it seems like ts2 should still be the timestamp of the insert (2), (4) transaction. :thinking:
___
*[`pkg/sql/logictest/testdata/logic_test/show_tenant_fingerprints` line 1 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqASAkQ4FmR_8ET3AwI:-NqASAkQ4FmR_8ET3AwJ:bmmei3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/show_tenant_fingerprints#L1)):*
> ```
> # LogicTest: !3node-tenant-default-configs !fakedist !fakedist-vec-off !fakedist-disk !local-read-committed
> ```

Why doesn't this one work?
___
*[`pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity` line 2 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAQqBgA623RBx53BsN:-NqAQqBgA623RBx53BsO:bb614s3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity#L2)):*
> ```
> # LogicTest: !local-read-committed
> # The local-read-committed configuration enables autocommit_before_ddl, which
> ```

Outdated comment?
___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 2 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAQyuh7udpa2LkBFg9:-NqAQyuh7udpa2LkBFgA:bb614s3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/txn#L2)):*
> ```
> # LogicTest: !local-read-committed
> # The local-read-committed config enables the autocommit_before_ddl, which
> ```

Outdated comment?
___
*[`pkg/sql/logictest/testdata/logic_test/txn_retry` line 1 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAR1JH17UQE5ilZycG:-NqAR1JH17UQE5ilZycH:b-ep8e6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/txn_retry#L1)):*
> ```
> # LogicTest: !local-read-committed
> ```

Would be nice to add a comment about retries.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
thanks for catching all of those skips! ready for another look

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-:-NqE0rdRCV_ZQgqJ5dMz:bdqrmnq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/schema_change_in_txn` line 26 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqARiWH43EWuW4xlRGa:-NqDtaAW-TycKh8_EN9g:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/ccl/logictestccl/testdata/logic_test/schema_change_in_txn#L26)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Maybe this should be a serializable transaction instead?
</blockquote></details>

fixed
___
*[`pkg/ccl/logictestccl/testdata/logic_test/udf_plpgsql` line 2866 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqARzDM55S-iSM48BHY:-NqDtunF9Wn6FvO5wQ9y:bqscnf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/ccl/logictestccl/testdata/logic_test/udf_plpgsql#L2866)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why is this one skipped?
</blockquote></details>

that was an oversight, nice catch
___
*[`pkg/sql/logictest/testdata/logic_test/fk` line 4175 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqA8qGh2a_3aWOh5Mzc:-NqDuCnQDSGVCFbkEvXN:bq3obn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/fk#L4175)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Does this table exist under RC?
</blockquote></details>

it does exist. could you expand on the question a bit?
___
*[`pkg/sql/logictest/testdata/logic_test/lock_timeout` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAFP590d2TncQl1Fyf:-NqDvDUMEG3OdV4HAn-l:bi6f2mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/lock_timeout#L3)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

What happens when this one is run under RC? Do lock timeouts not work with replicated locks?
</blockquote></details>

good question, i should have investigated if the differing behavior is expected or a bug with RC.

the error that happens is:
```
        .../pkg/sql/logictest/testdata/logic_test/lock_timeout:24:
        expected "canceling statement due to lock timeout on row \\(k\\)=\\(1\\) in t@t_pkey", but no error occurred
```

maybe @nvanbenschoten would have a clue if we'd expect that under read committed?
___
*[`pkg/sql/logictest/testdata/logic_test/manual_retry` line 1 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAHC4F5JHP8ynlkvOv:-NqDw64R94JbsSvKsymc:bxat96v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/manual_retry#L1)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Might be good to add a comment about automatically retrying under RC?
</blockquote></details>

thanks for the nudge. we actually can run this under RC, just need to look for a slightly different error message in one case
___
*[`pkg/sql/logictest/testdata/logic_test/read_committed` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAHwThEX2pViGI6Awu:-NqDwflx5dht_7YUixTK:b-jg6ewz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/read_committed#L3)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I guess the `autocommit_before_ddl` part is outdated now?
</blockquote></details>

nice catch
___
*[`pkg/sql/logictest/testdata/logic_test/schema_change_in_txn` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqALWR091PrcwZPwN8r:-NqDxRCm5jar2tmdg8GQ:b84lunq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn#L3)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Outdated comment.
</blockquote></details>

indeed, nice catch
___
*[`pkg/sql/logictest/testdata/logic_test/show_commit_timestamp` line 331 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAPNe44qqd6ANQBiP4:-NqDxe_50JqXtkCjw4KL:bsu89n2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/show_commit_timestamp#L331)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Hmm. Is `SHOW COMMIT TIMESTAMP` broken under RC? Oh, is it because the transaction isn't pushed? But it seems like ts2 should still be the timestamp of the insert (2), (4) transaction. :thinking:
</blockquote></details>

the error here is:
```
            /private/var/tmp/_bazel_rafiss/9796b4657e3f3e4bab52d5a958783c22/sandbox/darwin-sandbox/79514/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/logictest/tests/local-read-committed/local-read-committed_test_/local-read-committed_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/show_commit_timestamp:331: SELECT i,
                     CASE
                     WHEN ts = 1707500530525854000.0000000000 THEN 'ts1'
                     WHEN ts = 1707500530527876000.0000000000 THEN 'ts2'
                     END
                FROM (SELECT i, crdb_internal_mvcc_timestamp AS ts FROM foo)
            ORDER BY i ASC;
            expected:
                1  ts1
                2  ts2
                3  ts1
                4  ts2

            but found (query options: "") :
                1  ts1
                2  NULL
                3  ts1
                4  NULL
```

i may as well add an `onlyif read committed` version so that's more obvious. but also, i am unsure of the expected behavior here.
___
*[`pkg/sql/logictest/testdata/logic_test/show_tenant_fingerprints` line 1 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqASAkQ4FmR_8ET3AwI:-NqDyPE5AMZ7rVUhaFMk:b5ee7bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/show_tenant_fingerprints#L1)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why doesn't this one work?
</blockquote></details>

added a comment:
```
# This is skipped under local-read-committed, since it uses
# cluster_logical_timestamp() heavily, which does not work under
# READ COMMITTED.
```
___
*[`pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity` line 2 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAQqBgA623RBx53BsN:-NqDyXi321t0R5ZLWjov:b3dulwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity#L2)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Outdated comment?
</blockquote></details>

yeah, nice catch
___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 2 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAQyuh7udpa2LkBFg9:-NqE0_Z49W20bdl4nDdE:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/txn#L2)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Outdated comment?
</blockquote></details>

fixed
___
*[`pkg/sql/logictest/testdata/logic_test/txn_retry` line 1 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAR1JH17UQE5ilZycG:-NqE0gHdDzQHHFqPPN9X:b-w9a9fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/txn_retry#L1)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would be nice to add a comment about retries.
</blockquote></details>

rather than skipping the whole file, i skipped the one statement that doesn't work. (that statement seems to be the whole point of the test, but at least now if this file is augmented later, it will benefit from RC testing.)


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 4 of 30 files at r4, 116 of 116 files at r8, 1 of 1 files at r9, 118 of 130 files at r10, 1 of 1 files at r11, 11 of 11 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor` line 20 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqKleP2-iox6DEZc_K9:-NqKleP2-iox6DEZc_KA:b2lklme) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor#L20)):*
> ```
>   END
> $$ LANGUAGE PLpgSQL;
> BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;
> ```

Do we not support cursors in read committed transactions? Or is this because of the stored proc creation? Isn't that performed before this transaction begins?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor` line 814 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqKmPZ_-WziATg2dGdu:-NqKmPZ_-WziATg2dGdv:btmqt6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor#L814)):*
> ```
> <unnamed portal 10>
> 
> skipif config local-read-committed
> ```

Why do we need to skip here if we're also forcing the txn to run under serializable?
___
*[`pkg/sql/logictest/testdata/logic_test/alter_type` line 31 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqSxvJF1NzkJLS_EQrT:-NqSxvJF1NzkJLS_EQrU:b-w5zjyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/sql/logictest/testdata/logic_test/alter_type#L31)):*
> ```
> ROLLBACK
> 
> # We should be able to rename a type multiple types in a transaction.
> ```

`s/types/times/`?
___
*[`pkg/sql/logictest/testdata/logic_test/lock_timeout` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAFP590d2TncQl1Fyf:-NqSzv0EAmVYkslGw2K0:bso1w6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/lock_timeout#L3)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

good question, i should have investigated if the differing behavior is expected or a bug with RC.

the error that happens is:
```
        /private/var/tmp/_bazel_rafiss/9796b4657e3f3e4bab52d5a958783c22/sandbox/darwin-sandbox/78941/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/logictest/tests/local-read-committed/local-read-committed_test_/local-read-committed_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/lock_timeout:24:
        expected "canceling statement due to lock timeout on row \\(k\\)=\\(1\\) in t@t_pkey", but no error occurred
```

maybe @nvanbenschoten would have a clue if we'd expect that under read committed?
</blockquote></details>

I don't know which line that error is referring to, but I suspect the issue is that non-locking reads don't block on writes under RC, so there's no need for the lock timeout. So this would only be affecting the `SELECT`s that don't use `FOR UPDATE` or `FOR SHARE`.

It seems worthwhile to add a `skipif config local-read-committed` / `onlyif config local-read-committed` case for each of these.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 2 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqT093z6f_Dsw0ozZ_Y:-NqT094-7oeNK4bFq7Vv:bwzmqvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/sql/logictest/testdata/logic_test/select_for_share#L2)):*
> ```
> # LogicTest: !local-mixed-23.1 !local-read-committed
> # This test assumes that SERIALIZABLE isolation is being used.
> ```

Does it need to? It would be nice to have test coverage for FOR SHARE under RC. Would it help avoid differences between RC and SSI if we set `enable_durable_locking_for_serializable` in addition to `enable_shared_locking_for_serializable`?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/logictest/testdata/logic_test/lock_timeout` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAFP590d2TncQl1Fyf:-NqT4yuXAYr8dt9spRFs:bpt5r27) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/lock_timeout#L3)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't know which line that error is referring to, but I suspect the issue is that non-locking reads don't block on writes under RC, so there's no need for the lock timeout. So this would only be affecting the `SELECT`s that don't use `FOR UPDATE` or `FOR SHARE`.

It seems worthwhile to add a `skipif config local-read-committed` / `onlyif config local-read-committed` case for each of these.
</blockquote></details>

sorry, the line number is at the end of the file name: `lock_timeout:24`

making it assert each case explicitly sounds good to me


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
thanks for checking all my work / making sure nothing is unnecessarily skipped. RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-:-NqTi8VVDPx9DDUBPHXF:borix0v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor` line 20 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqKleP2-iox6DEZc_K9:-NqTWyQHBfQkIgePa0qR:bk4lpsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor#L20)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we not support cursors in read committed transactions? Or is this because of the stored proc creation? Isn't that performed before this transaction begins?
</blockquote></details>

nice catch, i used an overly aggressive find/replace. i've changed it so it only uses SERIALIZABLE when the CREATE FUNCTION schema change needs to happen in the same transaction as the one using the function.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor` line 814 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqKmPZ_-WziATg2dGdu:-NqTWXRb-eXTjOodBobe:b3wk3py) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/ccl/logictestccl/testdata/logic_test/plpgsql_cursor#L814)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we need to skip here if we're also forcing the txn to run under serializable?
</blockquote></details>

oops, this one was vestigial. removed this skip and the ones below
___
*[`pkg/sql/logictest/testdata/logic_test/alter_type` line 31 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqSxvJF1NzkJLS_EQrT:-NqTXLAf05QTOn-vYtwL:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/sql/logictest/testdata/logic_test/alter_type#L31)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/types/times/`?
</blockquote></details>

fixed
___
*[`pkg/sql/logictest/testdata/logic_test/lock_timeout` line 3 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqAFP590d2TncQl1Fyf:-NqTeUZi5NUH2R5iCHKR:brcgxc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4474af3fa18c9cecd630ef2de4ac311d226cdfda/pkg/sql/logictest/testdata/logic_test/lock_timeout#L3)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

sorry, the line number is at the end of the file name: `lock_timeout:24`

making it assert each case explicitly sounds good to me
</blockquote></details>

there were some SFU/UPDATE/DELETE cases that did not timeout under read committed. please see the latest patch and if there's any commentary that i can add for those cases, i'll do so
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 2 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-NqT093z6f_Dsw0ozZ_Y:-NqTfDTb2p73E-fPYH1r:bwvdd1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b67ab30f8e715a5bdc52d482b873ccd253d776/pkg/sql/logictest/testdata/logic_test/select_for_share#L2)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does it need to? It would be nice to have test coverage for FOR SHARE under RC. Would it help avoid differences between RC and SSI if we set `enable_durable_locking_for_serializable` in addition to `enable_shared_locking_for_serializable`?
</blockquote></details>

nice catch, this was also a vestigial skip from my first attempt at making this config. i've broken it down just to 3 cases that differ between RC/serializable


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

We should wait for a sign off from @michae2 as well.

> Reviewed 5 of 5 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-:-NqTjb-Z1K3UsKhNHs87:b-dh5prm)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work Rafi!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107990#-:-NqmM0Pd2NAlRkCV9CmV:bf2cpkk)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss commented:
tftrs!

bors r+
